### ISSUE TITLE
Pull the Solvers Out to Another Package?

### ISSUE THREAD

**Issue by ChrisRackauckas** [JuliaGNI/GeometricIntegrators.jl#2]
It seems you have a good set of linear and nonlinear solvers you're developing. What about pulling these out to separate packages, so that way one could use the linear solvers directly, or use the nonlinear solvers directly? I think those would be two very nice packages in the Julia ecosystem.

**Comment by michakraus** [comment]
Yes, I have been thinking about this myself, at least for the nonlinear solvers. So far there is not too much special about them, it's just Newton and quasi-Newton, but I plan to implement some more sophisticated versions, which could probably be useful elsewhere.

As for the linear solvers, these are just taken from Julia base and stripped of everything that requires dynamical allocation. So they are practically already there.

**Comment by ChrisRackauckas** [comment]
>As for the linear solvers, these are just taken from Julia base and stripped of everything that requires dynamical allocation. So they are practically already there.

I don't get this. The ones from Base don't use dynamical allocation if you use the `fact!` and `A_ldiv_B!` versions. How are the linear solvers different from those?

**Comment by michakraus** [comment]
As far as I understood the Julia Base code, when calling e.g. lufact!, this creates a LU object. I agree that this generates only very minor overhead, but these factors of 1.x add up quickly.

**Comment by ChrisRackauckas** [comment]
> when calling e.g. lufact!, this creates a LU object. I agree that this generates only very minor overhead, but these factors of 1.x add up quickly.

What do you mean factors of 1.x? You have the ~300KB or something like that from instantiating the type, but looking at your code you're instantiating a similar type, so I'm not sure what's gained. Usually when trying to avoid allocations like this (and those for things like very small pointers) you can actually lose performance. Can you show an example code of your version vs the Base version?

Note that this is also improved by using the `A_ldiv_B!` versions, which is easier via InplaceOps.jl:

```julia
lufact!(A)
@into! x   = A\b # inplace into x
@into! x2 = A\b # Re-uses the factorization so no factorization cost
```

**Comment by michakraus** [comment]
>> when calling e.g. lufact!, this creates a LU object. I agree that this generates only very minor overhead, but these factors of 1.x add up quickly.
>
>What do you mean factors of 1.x? You have the ~300KB or something like that from instantiating the type, but looking at your code you're instantiating a similar type, so I'm not sure what's gained. Usually when trying to avoid allocations like this (and those for things like very small pointers) you can actually lose performance. Can you show an example code of your version vs the Base version?

Yes, I have a similar type, but I am instantiating this exactly once vs. for each linear solve. I don't have a script for comparing the two at hand. I am relatively new to Julia and I am quite sure not everything I am doing in GeomDAE is the best way of doing it or even makes sense. However, most of the things I did in a somewhat unconventional way I prototyped in various different ways and kept the one that worked best (which usually means fastest). I did not keep all the "failures", though.

By factors of 1.x I mean that I implemented a few things the way I did in order to get a speedup of ~1.x with x potentially rather small. So each of these speedups on its own is not very meaningful, but having many of those little speedups actually results in something useful. I wouldn't necessarily do this in a code, but for a library it seems to make sense.

>Note that this is also improved by using the `A_ldiv_B!` versions, which is easier via InplaceOps.jl:

I think this wasn't available back when I looked at this. But the whole LU package from Base has evolved quite a bit since then. It's probably worth re-evaluating this, although not exactly a priority.

PS: Sorry, I am bit confused this morning (slightly sick actually). I hit the wrong button and accidentally changed your previous comment. I think I restored it, though.
